agent/eventhandler/handler_test.go

```diff
@@ -93,6 +95,35 @@ func TestSendsEventsOneEventRetries(t *testing.T) {
 	wg.Wait()
 }
 
+func TestSendsEventsInvalidParametersEventsRemoved(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	client := mock_api.NewMockECSClient(ctrl)
+	stateManager := statemanager.NewNoopStateManager()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	handler := NewTaskHandler(ctx, stateManager, nil, client)
+	defer cancel()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	taskEvent := taskEvent(taskARN)
+
+	client.EXPECT().SubmitTaskStateChange(gomock.Any()).Do(func(interface{}) {
+		assert.Equal(t, 1, handler.tasksToEvents[taskARN].events.Len())
+		wg.Done()
+	}).Return(awserr.New(ecs.ErrCodeInvalidParameterException, "", nil))
+
+	handler.AddStateChangeEvent(taskEvent, client)
+
+	wg.Wait()
+	// Require the lock to wait for submitFirstEvent to be finished
+	handler.tasksToEvents[taskARN].lock.Lock()
```

Comment:
There may be a race condition here, as the SubmitTaskStateChange is called in a goroutine, can you add a retry here?

Reply:
Can you please explain a bit why the race condition may happen?  I think `handler.tasksToEvents[taskARN].lock.Lock()` will not run until `SubmitTaskStateChange` is called as we have `wg.wait()`. I run the tests for 10000 times and they passed.

Reply:
See the following execution order as an example:
```
1. handler.AddStateChangeEvent(taskEvent, client)
2. SubmitTaskStateChange(this is in go routine)
3. wg.Done
4. wg.Wait
5. assert.Equal(t, 0, handler.tasksToEvents[taskARN].events.Len()) #this will fail, because it's not removed from the list
6. taskEvents.events.Remove(eventToSubmit)
```

Reply:
Hi, I used a little trick here, because before `SubmitTaskStateChange ` in `submitFirstEvent` is called,  `taskEvents.lock.Lock()` will be called first, and this lock will be unlocked when `submitFirstEvent` is finished. So before I do `assert.Equal(t, 0, handler.tasksToEvents[taskARN].events.Len())`, first I try to get the lock `handler.tasksToEvents[taskARN].lock.Lock()`, and we will block here until `submitFirstEvent` is finished, in which case we are guarantee that the event will be removed before the assertion (if the remove really happens).

Reply:
Oh, sorry, apparently I missed that.

Reply:
It's ok, it works in the same way as doing a retry.